src/use-swr-infinite.ts

```diff
@@ -0,0 +1,200 @@
+
+import defaultConfig, { cache } from './config'
+import SWRConfigContext from './swr-config-context'
+import useSWR from './use-swr'
+
+import { keyType, fetcherFn, ConfigInterface, responseInterface } from './types'
+type KeyLoader<Data = any> = (
+  index: number,
+  previousPageData: Data | null
+) => keyType
+type ExtendedConfigInterface<Data = any, Error = any> = ConfigInterface<
+  Data[],
+  Error,
+  fetcherFn<Data[]>
+> & {
+  initialSize?: number
+  revalidateAll?: boolean
+}
+type ExtendedResponseInterface<Data = any, Error = any> = responseInterface<
+  Data[],
+  Error
+> & {
+  size?: number
+  setSize?: (
+    size: number | ((size: number) => number)
+  ) => Promise<Data[] | undefined>
+}
+
+function useSWRInfinite<Data = any, Error = any>(
+  getKey: KeyLoader<Data>
+): ExtendedResponseInterface<Data, Error>
+function useSWRInfinite<Data = any, Error = any>(
+  getKey: KeyLoader<Data>,
+  config?: ExtendedConfigInterface<Data, Error>
+): ExtendedResponseInterface<Data, Error>
+function useSWRInfinite<Data = any, Error = any>(
+  getKey: KeyLoader<Data>,
+  fn?: fetcherFn<Data>,
+  config?: ExtendedConfigInterface<Data, Error>
+): ExtendedResponseInterface<Data, Error>
+function useSWRInfinite<Data = any, Error = any>(
+  ...args
+): ExtendedResponseInterface<Data, Error> {
+  let getKey: KeyLoader<Data>,
+    fn: fetcherFn<Data> | undefined,
+    config: ExtendedConfigInterface<Data, Error> = {}
+
+  if (args.length >= 1) {
+    getKey = args[0]
+  }
+  if (args.length > 2) {
+    fn = args[1]
+    config = args[2]
+  } else {
+    if (typeof args[1] === 'function') {
+      fn = args[1]
+    } else if (typeof args[1] === 'object') {
+      config = args[1]
+    }
+  }
+
+  config = Object.assign(
+    {},
+    defaultConfig,
+    useContext(SWRConfigContext),
+    config
+  )
+  let {
+    initialSize = 1,
+    revalidateAll = false,
+    fetcher: defaultFetcher,
+    ...extraConfig
+  } = config
+
+  if (typeof fn === 'undefined') {
+    // use the global fetcher
+    // we have to convert the type here
+    fn = (defaultFetcher as unknown) as fetcherFn<Data>
+  }
+
+  // get the serialized key of the first page
+  let firstPageKey: string | null = null
+  try {
+    ;[firstPageKey] = cache.serializeKey(getKey(0, null))
+  } catch (err) {
+    // not ready
+  }
+
+  const rerender = useState<boolean>(false)[1]
+
+  // we use cache to pass extra info (context) to fetcher so it can be globally shared
+  // here we get the key of the fetcher context cache
+  let contextCacheKey: string | null = null
+  if (firstPageKey) {
+    contextCacheKey = 'context@' + firstPageKey
+  }
+
+  // page count is cached as well, so when navigating the list can be restored
+  let pageCountCacheKey: string | null = null
+  if (firstPageKey) {
+    pageCountCacheKey = 'size@' + firstPageKey
+    initialSize = cache.get(pageCountCacheKey) || initialSize
+  }
+  const pageCountRef = useRef<number>(initialSize)
```

Comment:
It'd be nice if the page count updated automatically when the key changes, rather than requiring the user to reset the count to 0.

```javascript
if (firstPageKey) {
  pageCountCacheKey = 'size@' + firstPageKey
}
const pageCountRef = useRef<number>(initialSize)
useEffect(() => {
  pageCountRef.current = cache.get(pageCountCacheKey) || initialSize
}, [pageCountCacheKey, initialSize])
```

Reply:
Thanks for the feedback @neilgupta! Yeah I agree that in some cases it helps, but I don't think it should be the default behavior. Consider this scenario:

<img width="742" alt="Group 1" src="https://user-images.githubusercontent.com/3676859/85221542-07543d80-b3e7-11ea-843c-730d7c7a3fbe.png">

Say I have a list with a "show advanced" checkbox. When enabled the API should have a `?advanced=1` query attached. So when toggling the checkbox, I don't want the page count to be reset to 0. 

Therefore, if we make it the default behavior, it's gonna be hard for the user to bypass it. They'll have to memorize the page count before changing the key. So I think it's acceptable to manually reset the page size if we need to change the data source.

Reply:
An option might be nice for this, as I'm not sure loading more than one page when the query changes would be the behavior people expect, even if doing a `setSize` in an `useEffect` isn't that big of a deal, it could be a slight improvement

Reply:
Thanks again for the suggestion @neilgupta and @balthazar!

I thought more about this and decided to make it a default behavior but also provide an option. That said, page size will be reset automatically and you can also make it persistent. PR: https://github.com/vercel/swr/pull/490.